File: projects/Weighted Visit Assignment/backend/nano.py
"""The NANO study's participants, as the board's dropdowns need them.

The board's own data is coordinators, calendars and weights. The participants
and their anchor dates live in REDCap, and this is the join: it turns the
cached export into the two lists a scheduler picks from -- which family, and
which of that family's time points -- and hands the chosen window to the same
engine everything else uses.

Nothing here calls REDCap. The sync writes a cache and this reads it, so a
scheduling screen never depends on a study API being up, and the token is not
touched by anything a browser can reach.

WHAT IS AND IS NOT SENT TO A BROWSER. A participant ID and a visit window,
because choosing a visit is impossible without them. Never a date of birth or
a due date: those are what the windows were *derived from*, the derivation
happens here, and a screen that never receives them cannot leak them.
"""

from __future__ import annotations

from datetime import date
from typing import Dict, List, Optional

from esd_scheduler import redcap

#: The order a scheduler cares about. A window that has closed is the one
#: somebody needs to be told about; one three months out is not.
STATE_ORDER = {"missed": 0, "open": 1, "upcoming": 2, "done": 3}

STATE_LABEL = {
    "missed": "Window closed",
    "open": "Open now",
    "upcoming": "Not open yet",
    "done": "Already done",
    "no anchor date on file": "No anchor date",
}


def _families() -> List[redcap.NanoFamily]:
    cached = redcap.cached_families()
    out = []
    for row in cached.get("families") or []:
        out.append(redcap.NanoFamily(
            family_id=row.get("family_id", ""),
            participant_status=row.get("participant_status", "TD"),
            birth_date=row.get("birth_date"),
            due_date=row.get("due_date"),
            unenrolled=bool(row.get("unenrolled")),
            completed=list(row.get("completed") or []),
        ))
    return out


def summary(today: Optional[date] = None) -> dict:
    """Every family, with the one window that matters, for the first dropdown.

    Sorted by how urgent the next window is rather than by id: a list of two
    hundred participants in numerical order is a list nobody can act on.
    """
    today = today or date.today()
    cached = redcap.cached_families()
    rows = []
    for fam in _families():
        window = redcap.next_window(fam, today)
        rows.append({
            "family_id": fam.family_id,
            "participant_status": fam.participant_status,
            "completed_count": len(fam.completed),
            "next": window,
            "state": (window or {}).get("status", "none"),
            "state_label": STATE_LABEL.get((window or {}).get("status", ""), ""),
            # Deliberately no birth_date and no due_date. See the module note.
        })
    rows.sort(key=lambda r: (STATE_ORDER.get(r["state"], 9),
                             (r["next"] or {}).get("window_start") or "",
                             r["family_id"]))
    counts: Dict[str, int] = {}
    for row in rows:
        counts[row["state"]] = counts.get(row["state"], 0) + 1
    return {
        "fetched_at": cached.get("fetched_at"),
        "reason": cached.get("reason"),
        "count": len(rows),
        "counts_by_state": counts,
        "families": rows,
    }


def family_windows(family_id: str, today: Optional[date] = None) -> dict:
    """Every checkpoint for one family, for the second dropdown.

    Returned whole rather than filtered to the open one, because a scheduler
    catching up on a missed 9m needs to see it is missed, and one booking
    ahead needs to see what is not open yet. The states are labelled here so
    the page does not have to know the vocabulary.
    """
    today = today or date.today()
    fam = next((f for f in _families() if f.family_id == family_id), None)
    if fam is None:
        return {"family_id": family_id, "found": False, "windows": []}
    windows = redcap.visit_windows(fam, today)
    for window in windows:
        window["state_label"] = STATE_LABEL.get(window["status"], window["status"])
        window["selectable"] = bool(window["window_start"]) and not window["done"]
    return {
        "family_id": fam.family_id,
        "found": True,
        "participant_status": fam.participant_status,
        "completed": sorted(fam.completed, key=lambda c: int(
            "".join(ch for ch in c if ch.isdigit()) or 0)),
        "windows": windows,
    }

File: docs/visitboard/js/nano.js
/* ESD Visitboard - nano.js

   The NANO study's own screen: pick a family, pick a time point, get a pair.

   The rest of the board starts from a visit that already exists. This starts
   from the study: two hundred participants, each with eight checkpoints whose
   windows are worked out from their anchor date, and the question a
   coordinator actually arrives with is "who takes 5901's nine month visit".

   Two dropdowns and a button, in that order, because that is the order the
   question is asked in. The second dropdown cannot be filled until the first
   is answered -- a time point means nothing without a family -- so it stays
   disabled rather than showing a list that would be wrong.

   The prediction is not computed here. It is the same engine and the same
   mindmap the assign screen uses; this only chooses the window and hands it
   over. A second scoring path would drift from the first, and then two
   screens would disagree about who should take a visit.
*/

/* -----------------------------------------------------------------------
   SCREEN C  --  PICK A PARTICIPANT
  
     before  /api/nano/families answered with the study's participants
     here    two dropdowns in the order the question is asked: which
             family, then which of that family's eight time points
     after   assign.js shows the decision for the window that was chosen
  
     worked example
       filter chips: window closed 124, open now 36, not open yet 44
       the time-point dropdown stays disabled until a family is chosen:
         a time point means nothing without one
       'Who should take it?' -> POST /api/nano/plan -> the assign screen
   ----------------------------------------------------------------------- */

const NANO_STATE_ORDER = ["missed", "open", "upcoming", "done"];

function drawNano() {
  const card = $("nano-card");
  if (!card) return;
  api("/api/nano/families").then((data) => {
    S.nano = S.nano || {};
    S.nano.summary = data;
    renderNanoPickers(data);
  }).catch((err) => {
    $("nano-status").innerHTML =
      `<div class="notice notice-warn"><span>&#9432;</span><span>${
        esc(err.message)}</span></div>`;
  });
}

function renderNanoPickers(data) {
  const families = data.families || [];
  if (!data.fetched_at) {
    $("nano-status").innerHTML = `
      <div class="notice notice-warn"><span>&#9432;</span><span>
        No study export on file yet. Run <b>make redcap-sync</b> and reload.
        ${esc(data.reason || "")}</span></div>`;
    $("nano-pickers").innerHTML = "";
    return;
  }

  const counts = data.counts_by_state || {};
  const filter = (S.nano && S.nano.filter) || "all";
  /* The counts are the filter. A list of two hundred participants in id order
     is a list nobody can act on; "124 windows have closed" is where somebody
     starts, and clicking it is the obvious next move. */
  const chips = ["all", ...NANO_STATE_ORDER.filter((s) => counts[s])]
    .map((state) => `
      <button class="chip${filter === state ? " is-on" : ""}" type="button"
              data-nano-filter="${esc(state)}">
        ${state === "all" ? `All ${data.count}` : `${esc(NANO_LABEL[state] || state)} ${counts[state]}`}
      </button>`).join("");

  $("nano-status").innerHTML = `
    <div class="chipbar" role="group" aria-label="Filter families">${chips}</div>
    <p class="note">${data.count} participants, synced ${
      esc((data.fetched_at || "").replace("T", " "))}. Windows are worked out
      from each family's anchor date and the protocol schedule, not stored in
      the study record.</p>`;

  const shown = families.filter((f) => filter === "all" || f.state === filter);
  const chosen = (S.nano && S.nano.family) || "";

  $("nano-pickers").innerHTML = `
    <div class="nanopick">
      <label class="tweakrow">
        <span class="tweaklabel"><b>Family</b>
          <span class="note">${shown.length} shown${
            filter === "all" ? "" : `, filtered to ${esc(NANO_LABEL[filter] || filter)}`}</span></span>
        <select class="input select" id="nano-family">
          <option value="">Choose a participant&hellip;</option>
          ${shown.map((f) => `
            <option value="${esc(f.family_id)}" ${f.family_id === chosen ? "selected" : ""}>
              ${esc(f.family_id)} &middot; ${esc(f.participant_status)}${
                f.next ? ` &middot; ${esc(f.next.checkpoint)} ${esc(f.state_label)}` : ""}
            </option>`).join("")}
        </select>
      </label>
      <label class="tweakrow">
        <span class="tweaklabel"><b>Time point</b>
          <span class="note">Which of this family's eight NANO checkpoints</span></span>
        <select class="input select" id="nano-checkpoint" disabled>
          <option value="">Pick a family first</option>
        </select>
      </label>
    </div>
    <div id="nano-window"></div>`;

  document.querySelectorAll("[data-nano-filter]").forEach((b) =>
    b.addEventListener("click", () => {
      S.nano.filter = b.dataset.nanoFilter;
      S.nano.family = "";
      renderNanoPickers(S.nano.summary);
    }));
  $("nano-family").addEventListener("change", (e) => chooseNanoFamily(e.target.value));
  if (chosen) chooseNanoFamily(chosen);
}

const NANO_LABEL = {
  missed: "Window closed", open: "Open now",
  upcoming: "Not open yet", done: "Done",
};

async function chooseNanoFamily(familyId) {
  S.nano = S.nano || {};
  S.nano.family = familyId;
  const picker = $("nano-checkpoint");
  if (!familyId) {
    picker.disabled = true;
    picker.innerHTML = `<option value="">Pick a family first</option>`;
    $("nano-window").innerHTML = "";
    return;
  }
  const data = await api("/api/nano/family?id=" + encodeURIComponent(familyId));
  S.nano.windows = data;
  const windows = data.windows || [];
  picker.disabled = false;
  picker.innerHTML = `<option value="">Choose a time point&hellip;</option>` +
    windows.map((w) => `
      <option value="${esc(w.checkpoint)}" ${w.selectable ? "" : "disabled"}>
        ${esc(w.checkpoint)} &middot; ${esc(w.state_label)}${
          w.window_start ? ` &middot; ${esc(w.window_start)} to ${esc(w.window_end)}` : ""}
      </option>`).join("");
  picker.onchange = (e) => chooseNanoWindow(e.target.value);
  $("nano-window").innerHTML = `
    <p class="note">${esc(familyId)} is <b>${esc(data.participant_status)}</b>.
      ${data.completed.length
        ? `Completed: ${data.completed.map(esc).join(", ")}.`
        : "No checkpoint recorded as done yet."}
      A preterm participant's windows up to 24m count from the due date; every
      36m visit counts from the birthday.</p>`;
}

async function chooseNanoWindow(checkpoint) {
  if (!checkpoint) { $("nano-window").innerHTML = ""; return; }
  const familyId = S.nano.family;
  const window_ = (S.nano.windows.windows || [])
    .find((w) => w.checkpoint === checkpoint);
  $("nano-window").innerHTML = `
    <div class="nanowindow">
      <div>
        <p class="eyebrow">${esc(familyId)} &middot; ${esc(checkpoint)}</p>
        <h3>${esc(window_.window_start)} to ${esc(window_.window_end)}</h3>
        <p class="note">Ideal date ${esc(window_.ideal)} &middot; ${
          esc(window_.state_label)}${window_.remote
            ? " &middot; remote checkpoint, nobody travels" : ""}</p>
      </div>
      <button class="btn btn-primary" type="button" id="nano-go">
        Who should take it?</button>
    </div>
    <div id="nano-result"></div>`;
  $("nano-go").addEventListener("click", () => planNanoVisit(familyId, checkpoint));
}

async function planNanoVisit(familyId, checkpoint) {
  const button = $("nano-go");
  if (button) { button.disabled = true; button.textContent = "Working it out…"; }
  try {
    const out = await api("/api/nano/plan", {
      method: "POST",
      body: JSON.stringify({ family_id: familyId, checkpoint }),
    });
    /* Hand over to the assign screen rather than drawing a second tree here.

       Rendering the mindmap in this section too put two of them in the
       document with the same element ids, so clicking a branch bound to
       whichever the browser found first and the other went stale. One tree,
       one place: this screen's job is choosing the window, and the decision
       has a screen already. */
    S.selected = out.visit_id;
    S.detail = out.detail;
    resetMindmap();
    await refresh();
    setSection("assign");
    await selectVisit(out.visit_id);
    toast(`${familyId} ${checkpoint} is on the board.`);
  } catch (err) {
    $("nano-result").innerHTML =
      `<div class="notice notice-alert"><span>&#9888;</span><span>${
        esc(err.message)}</span></div>`;
  } finally {
    if (button) { button.disabled = false; button.textContent = "Who should take it?"; }
  }
}
